File: Mathematics/Fast_Exponentiation.py
# This is the program to calculate the exponents much faster in O(log n base 2) time

# Function to calculate the exponent of the number 
def power(num,expo):
    ''' BASE CASE '''
    if expo==0:
        return 1

    ''' RECURSSIVE CASE '''
    ans=power(num,expo//2)
    ans=ans*ans

    ''' IF THE POWER IS ODD THEN MULTIPLY NUMBER ONE MORE TIME ''' 
    if (expo&1):
        ans=ans*num
    return ans

num,expo=map(int,input("Enter the number and the exponent\n").split())
ans=power(num,expo)
print(ans)

File: 04-Mathematics/Fast_Exponentiation.cpp
//
// Created by soumil on 8/7/2019.
//
#include<iostream>
using namespace std;

int power(int number,int exponent)
{
    // base case
    if(exponent==0)
    {
        return 1;
    }

    // Recursive case
    int ans =power(number,exponent/2);
    ans *= ans;

    if (exponent&1)
    {
        ans=ans*number;
    }
    return ans;
}
int main()
{
    int number, exponent;
    cout<<"Enter the number and the exponent"<<endl;
    cin>>number>>exponent;
    cout<<power(number,exponent);
    return 0;
}
